### PR TITLE
Weight structured sparsity in training

### PR DESCRIPTION
**Pull request by BirdsOfAFthr**
# Description

- Enable Structured Sparsity in Qwix:
  - Added `get_fp8_full_qwix_rule_w_sparsity` in `quantizations.py` to configure N:M sparsity (N, M, and update steps) via `additional_qt_config`.
  
- Support Non-Differentiable State (`batch_stats`):
  - Updated `loss_fn`, `train_step`, and `eval_step` in `train.py` to accept and pass `batch_stats`.
  - Wrapped model application variables to include `{"params": params, "batch_stats": sparsity_state}`.
  - Ensured `train_step` filters gradients so that they are only applied to differentiable parameters, while `batch_stats` are updated from the model's auxiliary outputs.

- Safe Partial Checkpoint Loading:
  - Updated `setup_initial_state` in `maxtext_utils.py` to intelligently merge loaded parameters. It now ignores `jax.ShapeDtypeStruct` placeholders (which indicate un-loaded variables) and falls back to initialized parameters.

# Tests
* Added internal unit tests (will be in the G3 change)
* E2E benchmarking

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests tests and provided workload links above if applicable.
- [X] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/configs/base.yml [hunk 1]
```diff
@@ -149,6 +149,14 @@ act_quantization_calibration_method: "absmax"
 bwd_quantization_calibration_method: "absmax"
 # shard the range finding operation for quantization. by default this is set to number of slices.
 quantization_local_shard_count: -1
+# The 'N' in N:M sparsity, representing the maximum number of non-zero values in each block.
+weight_sparsity_n: null
+# The 'M' in N:M sparsity, representing the number of values in each block.
+weight_sparsity_m: null
```
Comment (shuningjin) [inline src/maxtext/configs/base.yml:155]: nit: null -> None, to match maxtext config style.

File: src/dependencies/requirements/generated_requirements/tpu-requirements.txt [hunk 2]
```diff
@@ -178,7 +178,7 @@ python-dateutil>=2.9.0.post0
 pytokens>=0.4.1
 pytype>=2024.10.11
 pyyaml>=6.0.3
-qwix>=0.1.5
+qwix>=0.1.6
```
Comment (shuningjin) [inline src/dependencies/requirements/generated_requirements/tpu-requirements.txt:181]: Need to check qwix update with team
- Seems our unit tests get broken - [test log](https://github.com/AI-Hypercomputer/maxtext/actions/runs/24704319723/job/72451065141?pr=3561), due to `from qwix.contrib.sparsity import sparsity_module`,
- Need to update qwix in our github runner, otherwise might block other PRs

**Review by shuningjin — COMMENTED** [review]
Thank you for the feature! LGTM. Need to check dependency upgrade on our end.

File: src/dependencies/requirements/generated_requirements/tpu-requirements.txt (hunk 2 above)
Comment (SurbhiJainUSC) [reply src/dependencies/requirements/generated_requirements/tpu-requirements.txt:181]: Please don't edit this file manually: https://screenshot.googleplex.com/8BSMDqDioChJjHE

Guide to update dependencies: https://maxtext.readthedocs.io/en/latest/development/update_dependencies.html

Comment (BirdsOfAFthr) [reply src/dependencies/requirements/generated_requirements/tpu-requirements.txt:181]: I followed that [guide](https://maxtext.readthedocs.io/en/latest/development/update_dependencies.html).

Do you know which commit to choose? No matter what I choose from [JAX build/ folder](https://github.com/jax-ml/jax/commits/main/build), the unit test is still failing.

**Review by shuningjin — APPROVED** [review]
